Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -239,21 +250,257 @@ FieldValue Serializer::DecodeFieldValue(const uint8_t* bytes, size_t length) {
+  // Implementation note: This is roughly modeled on pb_decode_delimited,
+  // adjusted to account for the oneof in FieldValue.
+  pb_istream_t substream;
+  bool status = pb_make_string_substream(stream, &substream);
+  if (!status) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+
+  FieldValue fv = DecodeFieldValueImpl(&substream);
+
+  // NB: future versions of nanopb read the remaining characters out of the
+  // substream (and return false if that fails) as an additional safety
+  // check within pb_clost_string_substream. Unfortunately, that's not present
+  // in the current version (0.38).  We'll make a stronger assertion and check
+  // to make sure there *are* no remaining characters in the substream.
+  if (substream.bytes_left != 0) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+  pb_close_string_substream(stream, &substream);
+
+  return fv;
+}
+
+void EncodeFieldEntry(
+    pb_ostream_t* stream,
+    const std::pair<const std::string, const FieldValue>& kv) {
+  // Calculate the size of this FieldEntry. This is the size of the key and
+  // value, plus an additional 2 for the two tags.
+  pb_ostream_t sizing_stream = PB_OSTREAM_SIZING;
+  EncodeString(&sizing_stream, kv.first);
+  EncodeSubFieldValue(&sizing_stream, kv.second);
+
+  // additional 2 for the two tags
+  EncodeVarint(stream, sizing_stream.bytes_written + 2);
+
+  // Encode the key (string)
+  bool status =
+      pb_encode_tag(stream, PB_WT_STRING,
+                    google_firestore_v1beta1_MapValue_FieldsEntry_key_tag);
+  if (!status) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+  EncodeString(stream, kv.first);
+
+  // Encode the value (FieldValue)
+  status =
+      pb_encode_tag(stream, PB_WT_STRING,
+                    google_firestore_v1beta1_MapValue_FieldsEntry_value_tag);
+  if (!status) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+  EncodeSubFieldValue(stream, kv.second);
+}
+
+std::pair<const std::string, const FieldValue> DecodeFieldEntry(
+    pb_istream_t* stream) {
+  pb_wire_type_t wire_type;
+  uint32_t tag;
+  bool eof;
+  bool status = pb_decode_tag(stream, &wire_type, &tag, &eof);
+  // TODO(rsgowman): figure out error handling: We can do better than a failed
+  // assertion.
+  FIREBASE_ASSERT(tag == google_firestore_v1beta1_MapValue_FieldsEntry_key_tag);
+  FIREBASE_ASSERT(wire_type == PB_WT_STRING);
+  FIREBASE_ASSERT(!eof);
+  FIREBASE_ASSERT(status);
+  std::string key = DecodeString(stream);
+
+  status = pb_decode_tag(stream, &wire_type, &tag, &eof);
+  FIREBASE_ASSERT(tag ==
+                  google_firestore_v1beta1_MapValue_FieldsEntry_value_tag);
+  // NB: PB_WT_STRING is used for submessages too.
+  FIREBASE_ASSERT(wire_type == PB_WT_STRING);
+  FIREBASE_ASSERT(!eof);
+  FIREBASE_ASSERT(status);
+
+  FieldValue value = DecodeSubFieldValue(stream);
+
+  return std::make_pair(key, value);
```

Comment:
Nit: I think you can just `return {key, value};`.

Reply:
done.